{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# KEGG API Biopython Notebook\n",
    "\n",
    "This notebook features basic functions from biopython.kegg module that facilitate the use\n",
    "of KEGG's API. \n",
    "\n",
    "With these basic functions you are able to search, find and recover info from KEGG database \n",
    "that are very important in a diversity of analysis.\n",
    "\n",
    "\n",
    "\n",
    "## Install necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install pandas"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load packages into notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import REST\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load function\n",
    "\n",
    "This function is going to be used below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_df(data):\n",
    "    lines = data.read().strip().split(\"\\n\")\n",
    "    data_list = []\n",
    "    for line in lines:\n",
    "        parts = line.split(\"\\t\")\n",
    "        if len(parts) == 2:\n",
    "            data_list.append(parts)          \n",
    "    return pd.DataFrame(data_list, columns=[\"Data\", \"Name\"])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explore KEGG databases\n",
    "\n",
    "KEGG is a big database made of subdatabases. To list all the databases available you can use the function `kegg_info` \n",
    "with \"kegg\". \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kegg             Kyoto Encyclopedia of Genes and Genomes\n",
      "kegg             Release 110.0+/06-08, Jun 24\n",
      "                 Kanehisa Laboratories\n",
      "                 pathway   1,172,423 entries\n",
      "                 brite       390,942 entries\n",
      "                 module          561 entries\n",
      "                 orthology    26,794 entries\n",
      "                 genome       24,739 entries\n",
      "                 genes     54,564,849 entries\n",
      "                 compound     19,356 entries\n",
      "                 glycan       11,220 entries\n",
      "                 reaction     12,088 entries\n",
      "                 rclass        3,194 entries\n",
      "                 enzyme        8,158 entries\n",
      "                 network       1,549 entries\n",
      "                 variant       1,452 entries\n",
      "                 disease       2,750 entries\n",
      "                 drug         12,449 entries\n",
      "                 dgroup        2,471 entries\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(REST.kegg_info(\"kegg\").read())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To explore a specific database, use its name in the function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "module           KEGG Module Database\n",
      "md               Release 110.0+/06-08, Jun 24\n",
      "                 Kanehisa Laboratories\n",
      "                 561 entries\n",
      "\n",
      "linked db        pathway\n",
      "                 ko\n",
      "                 <org>\n",
      "                 genome\n",
      "                 compound\n",
      "                 glycan\n",
      "                 reaction\n",
      "                 enzyme\n",
      "                 pubmed\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(REST.kegg_info(\"module\").read())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Listing all data present in specific database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "map01100\tMetabolic pathways\n",
      "map01110\tBiosynthesis of secondary metabolites\n",
      "map01120\tMicrobial metabolism in diverse environments\n",
      "map01200\tCarbon metabolism\n",
      "map01210\t2-Oxocarboxylic acid metabolism\n",
      "map01212\tFatty acid metabolism\n",
      "map01230\tBiosynthesis of amino acids\n",
      "map01232\tNucleotide metabolism\n",
      "map01250\tBiosynthesis of nucleotide sugars\n",
      "map01240\tBiosynthesis of cofactors\n",
      "map01220\tDegradation of aromatic compounds\n",
      "map00010\tGlycolysis / Gluconeogenesis\n",
      "map00020\tCitrate cycle (TCA cycle)\n",
      "map00030\tPentose phosphate pathway\n",
      "map00040\tPentose and glucuronate interconversions\n",
      "map00051\tFructose and mannose metabolism\n",
      "map00052\tGalactose metabolism\n",
      "map00053\tAscorbate and aldarate metabolism\n",
      "map00500\tStarch and sucrose metabolism\n",
      "map00520\tAmino sugar and nucleotide sugar metabolism\n",
      "map00620\tPyruvate metabolism\n",
      "map00630\tGlyoxylate and dicarboxylate metabolism\n",
      "map00640\tPropanoate metabolism\n",
      "map00650\tButanoate metabolism\n",
      "map00660\tC5-Branched dibasic acid metabolism\n",
      "map00562\tInositol phosphate metabolism\n",
      "map00190\tOxidative phosphorylation\n",
      "map00195\tPhotosynthesis\n",
      "map00196\tPhotosynthesis - antenna proteins\n",
      "map00710\tCarbon fixation in photosynthetic organisms\n",
      "map00720\tCarbon fixation pathways in prokaryotes\n",
      "map00680\tMethane metabolism\n",
      "map00910\tNitrogen metabolism\n",
      "map00920\tSulfur metabolism\n",
      "map00061\tFatty acid biosynthesis\n",
      "map00062\tFatty acid elongation\n",
      "map00071\tFatty acid degradation\n",
      "map00073\tCutin, suberine and wax biosynthesis\n",
      "map00100\tSteroid biosynthesis\n",
      "map00120\tPrimary bile acid biosynthesis\n",
      "map00121\tSecondary bile acid biosynthesis\n",
      "map00140\tSteroid hormone biosynthesis\n",
      "map00561\tGlycerolipid metabolism\n",
      "map00564\tGlycerophospholipid metabolism\n",
      "map00565\tEther lipid metabolism\n",
      "map00600\tSphingolipid metabolism\n",
      "map00590\tArachidonic acid metabolism\n",
      "map00591\tLinoleic acid metabolism\n",
      "map00592\talpha-Linolenic acid metabolism\n",
      "map01040\tBiosynthesis of unsaturated fatty acids\n",
      "map00230\tPurine metabolism\n",
      "map00240\tPyrimidine metabolism\n",
      "map00250\tAlanine, aspartate and glutamate metabolism\n",
      "map00260\tGlycine, serine and threonine metabolism\n",
      "map00270\tCysteine and methionine metabolism\n",
      "map00280\tValine, leucine and isoleucine degradation\n",
      "map00290\tValine, leucine and isoleucine biosynthesis\n",
      "map00300\tLysine biosynthesis\n",
      "map00310\tLysine degradation\n",
      "map00220\tArginine biosynthesis\n",
      "map00330\tArginine and proline metabolism\n",
      "map00340\tHistidine metabolism\n",
      "map00350\tTyrosine metabolism\n",
      "map00360\tPhenylalanine metabolism\n",
      "map00380\tTryptophan metabolism\n",
      "map00400\tPhenylalanine, tyrosine and tryptophan biosynthesis\n",
      "map00410\tbeta-Alanine metabolism\n",
      "map00430\tTaurine and hypotaurine metabolism\n",
      "map00440\tPhosphonate and phosphinate metabolism\n",
      "map00450\tSelenocompound metabolism\n",
      "map00460\tCyanoamino acid metabolism\n",
      "map00470\tD-Amino acid metabolism\n",
      "map00480\tGlutathione metabolism\n",
      "map00510\tN-Glycan biosynthesis\n",
      "map00513\tVarious types of N-glycan biosynthesis\n",
      "map00512\tMucin type O-glycan biosynthesis\n",
      "map00515\tMannose type O-glycan biosynthesis\n",
      "map00514\tOther types of O-glycan biosynthesis\n",
      "map00532\tGlycosaminoglycan biosynthesis - chondroitin sulfate / dermatan sulfate\n",
      "map00534\tGlycosaminoglycan biosynthesis - heparan sulfate / heparin\n",
      "map00533\tGlycosaminoglycan biosynthesis - keratan sulfate\n",
      "map00531\tGlycosaminoglycan degradation\n",
      "map00563\tGlycosylphosphatidylinositol (GPI)-anchor biosynthesis\n",
      "map00601\tGlycosphingolipid biosynthesis - lacto and neolacto series\n",
      "map00603\tGlycosphingolipid biosynthesis - globo and isoglobo series\n",
      "map00604\tGlycosphingolipid biosynthesis - ganglio series\n",
      "map00511\tOther glycan degradation\n",
      "map00540\tLipopolysaccharide biosynthesis\n",
      "map00542\tO-Antigen repeat unit biosynthesis\n",
      "map00541\tO-Antigen nucleotide sugar biosynthesis\n",
      "map00550\tPeptidoglycan biosynthesis\n",
      "map00552\tTeichoic acid biosynthesis\n",
      "map00571\tLipoarabinomannan (LAM) biosynthesis\n",
      "map00572\tArabinogalactan biosynthesis - Mycobacterium\n",
      "map00543\tExopolysaccharide biosynthesis\n",
      "map00730\tThiamine metabolism\n",
      "map00740\tRiboflavin metabolism\n",
      "map00750\tVitamin B6 metabolism\n",
      "map00760\tNicotinate and nicotinamide metabolism\n",
      "map00770\tPantothenate and CoA biosynthesis\n",
      "map00780\tBiotin metabolism\n",
      "map00785\tLipoic acid metabolism\n",
      "map00790\tFolate biosynthesis\n",
      "map00670\tOne carbon pool by folate\n",
      "map00830\tRetinol metabolism\n",
      "map00860\tPorphyrin metabolism\n",
      "map00130\tUbiquinone and other terpenoid-quinone biosynthesis\n",
      "map00900\tTerpenoid backbone biosynthesis\n",
      "map00902\tMonoterpenoid biosynthesis\n",
      "map00909\tSesquiterpenoid and triterpenoid biosynthesis\n",
      "map00904\tDiterpenoid biosynthesis\n",
      "map00906\tCarotenoid biosynthesis\n",
      "map00905\tBrassinosteroid biosynthesis\n",
      "map00981\tInsect hormone biosynthesis\n",
      "map00908\tZeatin biosynthesis\n",
      "map00903\tLimonene degradation\n",
      "map00907\tPinene, camphor and geraniol degradation\n",
      "map01052\tType I polyketide structures\n",
      "map00522\tBiosynthesis of 12-, 14- and 16-membered macrolides\n",
      "map01051\tBiosynthesis of ansamycins\n",
      "map01059\tBiosynthesis of enediyne antibiotics\n",
      "map01056\tBiosynthesis of type II polyketide backbone\n",
      "map01057\tBiosynthesis of type II polyketide products\n",
      "map00253\tTetracycline biosynthesis\n",
      "map00523\tPolyketide sugar unit biosynthesis\n",
      "map01054\tNonribosomal peptide structures\n",
      "map01053\tBiosynthesis of siderophore group nonribosomal peptides\n",
      "map01055\tBiosynthesis of vancomycin group antibiotics\n",
      "map00940\tPhenylpropanoid biosynthesis\n",
      "map00945\tStilbenoid, diarylheptanoid and gingerol biosynthesis\n",
      "map00941\tFlavonoid biosynthesis\n",
      "map00944\tFlavone and flavonol biosynthesis\n",
      "map00942\tAnthocyanin biosynthesis\n",
      "map00943\tIsoflavonoid biosynthesis\n",
      "map00946\tDegradation of flavonoids\n",
      "map00901\tIndole alkaloid biosynthesis\n",
      "map00403\tIndole diterpene alkaloid biosynthesis\n",
      "map00950\tIsoquinoline alkaloid biosynthesis\n",
      "map00960\tTropane, piperidine and pyridine alkaloid biosynthesis\n",
      "map00996\tBiosynthesis of various alkaloids\n",
      "map00232\tCaffeine metabolism\n",
      "map00965\tBetalain biosynthesis\n",
      "map00966\tGlucosinolate biosynthesis\n",
      "map00402\tBenzoxazinoid biosynthesis\n",
      "map00311\tPenicillin and cephalosporin biosynthesis\n",
      "map00332\tCarbapenem biosynthesis\n",
      "map00261\tMonobactam biosynthesis\n",
      "map00331\tClavulanic acid biosynthesis\n",
      "map00521\tStreptomycin biosynthesis\n",
      "map00524\tNeomycin, kanamycin and gentamicin biosynthesis\n",
      "map00525\tAcarbose and validamycin biosynthesis\n",
      "map00401\tNovobiocin biosynthesis\n",
      "map00404\tStaurosporine biosynthesis\n",
      "map00405\tPhenazine biosynthesis\n",
      "map00333\tProdigiosin biosynthesis\n",
      "map00254\tAflatoxin biosynthesis\n",
      "map00998\tBiosynthesis of various antibiotics\n",
      "map00999\tBiosynthesis of various plant secondary metabolites\n",
      "map00997\tBiosynthesis of various other secondary metabolites\n",
      "map00362\tBenzoate degradation\n",
      "map00627\tAminobenzoate degradation\n",
      "map00364\tFluorobenzoate degradation\n",
      "map00625\tChloroalkane and chloroalkene degradation\n",
      "map00361\tChlorocyclohexane and chlorobenzene degradation\n",
      "map00623\tToluene degradation\n",
      "map00622\tXylene degradation\n",
      "map00633\tNitrotoluene degradation\n",
      "map00642\tEthylbenzene degradation\n",
      "map00643\tStyrene degradation\n",
      "map00791\tAtrazine degradation\n",
      "map00930\tCaprolactam degradation\n",
      "map00363\tBisphenol degradation\n",
      "map00621\tDioxin degradation\n",
      "map00626\tNaphthalene degradation\n",
      "map00624\tPolycyclic aromatic hydrocarbon degradation\n",
      "map00365\tFurfural degradation\n",
      "map00984\tSteroid degradation\n",
      "map00980\tMetabolism of xenobiotics by cytochrome P450\n",
      "map00982\tDrug metabolism - cytochrome P450\n",
      "map00983\tDrug metabolism - other enzymes\n",
      "map01010\tOverview of biosynthetic pathways\n",
      "map01060\tBiosynthesis of plant secondary metabolites\n",
      "map01061\tBiosynthesis of phenylpropanoids\n",
      "map01062\tBiosynthesis of terpenoids and steroids\n",
      "map01063\tBiosynthesis of alkaloids derived from shikimate pathway\n",
      "map01064\tBiosynthesis of alkaloids derived from ornithine, lysine and nicotinic acid\n",
      "map01065\tBiosynthesis of alkaloids derived from histidine and purine\n",
      "map01066\tBiosynthesis of alkaloids derived from terpenoid and polyketide\n",
      "map01070\tBiosynthesis of plant hormones\n",
      "map03020\tRNA polymerase\n",
      "map03022\tBasal transcription factors\n",
      "map03040\tSpliceosome\n",
      "map03010\tRibosome\n",
      "map00970\tAminoacyl-tRNA biosynthesis\n",
      "map03013\tNucleocytoplasmic transport\n",
      "map03015\tmRNA surveillance pathway\n",
      "map03008\tRibosome biogenesis in eukaryotes\n",
      "map03060\tProtein export\n",
      "map04141\tProtein processing in endoplasmic reticulum\n",
      "map04130\tSNARE interactions in vesicular transport\n",
      "map04120\tUbiquitin mediated proteolysis\n",
      "map04122\tSulfur relay system\n",
      "map03050\tProteasome\n",
      "map03018\tRNA degradation\n",
      "map03030\tDNA replication\n",
      "map03410\tBase excision repair\n",
      "map03420\tNucleotide excision repair\n",
      "map03430\tMismatch repair\n",
      "map03440\tHomologous recombination\n",
      "map03450\tNon-homologous end-joining\n",
      "map03460\tFanconi anemia pathway\n",
      "map03082\tATP-dependent chromatin remodeling\n",
      "map03083\tPolycomb repressive complex\n",
      "map03230\tViral genome structure\n",
      "map03240\tViral replication\n",
      "map03250\tViral life cycle - HIV-1\n",
      "map03260\tVirion - Human immunodeficiency virus\n",
      "map03271\tVirion - Rotavirus\n",
      "map03262\tVirion - Coronavirus\n",
      "map03261\tVirion - Influenza virus\n",
      "map03263\tVirion - Enterovirus\n",
      "map03264\tVirion - Flavivirus\n",
      "map03265\tVirion - Ebolavirus, Lyssavirus and Morbillivirus\n",
      "map03266\tVirion - Herpesvirus\n",
      "map03268\tVirion - Orthopoxvirus\n",
      "map03267\tVirion - Adenovirus\n",
      "map03259\tVirion - Bacteriophages T4 and T7\n",
      "map03258\tVirion - Bacteriophage lambda\n",
      "map02010\tABC transporters\n",
      "map02060\tPhosphotransferase system (PTS)\n",
      "map03070\tBacterial secretion system\n",
      "map02020\tTwo-component system\n",
      "map04010\tMAPK signaling pathway\n",
      "map04013\tMAPK signaling pathway - fly\n",
      "map04016\tMAPK signaling pathway - plant\n",
      "map04011\tMAPK signaling pathway - yeast\n",
      "map04012\tErbB signaling pathway\n",
      "map04014\tRas signaling pathway\n",
      "map04015\tRap1 signaling pathway\n",
      "map04310\tWnt signaling pathway\n",
      "map04330\tNotch signaling pathway\n",
      "map04340\tHedgehog signaling pathway\n",
      "map04341\tHedgehog signaling pathway - fly\n",
      "map04350\tTGF-beta signaling pathway\n",
      "map04390\tHippo signaling pathway\n",
      "map04391\tHippo signaling pathway - fly\n",
      "map04392\tHippo signaling pathway - multiple species\n",
      "map04370\tVEGF signaling pathway\n",
      "map04371\tApelin signaling pathway\n",
      "map04630\tJAK-STAT signaling pathway\n",
      "map04064\tNF-kappa B signaling pathway\n",
      "map04668\tTNF signaling pathway\n",
      "map04066\tHIF-1 signaling pathway\n",
      "map04068\tFoxO signaling pathway\n",
      "map04020\tCalcium signaling pathway\n",
      "map04070\tPhosphatidylinositol signaling system\n",
      "map04072\tPhospholipase D signaling pathway\n",
      "map04071\tSphingolipid signaling pathway\n",
      "map04024\tcAMP signaling pathway\n",
      "map04022\tcGMP-PKG signaling pathway\n",
      "map04151\tPI3K-Akt signaling pathway\n",
      "map04152\tAMPK signaling pathway\n",
      "map04150\tmTOR signaling pathway\n",
      "map04075\tPlant hormone signal transduction\n",
      "map04080\tNeuroactive ligand-receptor interaction\n",
      "map04060\tCytokine-cytokine receptor interaction\n",
      "map04061\tViral protein interaction with cytokine and cytokine receptor\n",
      "map04512\tECM-receptor interaction\n",
      "map04514\tCell adhesion molecules\n",
      "map04144\tEndocytosis\n",
      "map04145\tPhagosome\n",
      "map04142\tLysosome\n",
      "map04146\tPeroxisome\n",
      "map04140\tAutophagy - animal\n",
      "map04138\tAutophagy - yeast\n",
      "map04136\tAutophagy - other\n",
      "map04137\tMitophagy - animal\n",
      "map04139\tMitophagy - yeast\n",
      "map04148\tEfferocytosis\n",
      "map04110\tCell cycle\n",
      "map04111\tCell cycle - yeast\n",
      "map04112\tCell cycle - Caulobacter\n",
      "map04113\tMeiosis - yeast\n",
      "map04114\tOocyte meiosis\n",
      "map04210\tApoptosis\n",
      "map04214\tApoptosis - fly\n",
      "map04215\tApoptosis - multiple species\n",
      "map04216\tFerroptosis\n",
      "map04217\tNecroptosis\n",
      "map04115\tp53 signaling pathway\n",
      "map04218\tCellular senescence\n",
      "map04510\tFocal adhesion\n",
      "map04520\tAdherens junction\n",
      "map04530\tTight junction\n",
      "map04540\tGap junction\n",
      "map04550\tSignaling pathways regulating pluripotency of stem cells\n",
      "map02024\tQuorum sensing\n",
      "map05111\tBiofilm formation - Vibrio cholerae\n",
      "map02025\tBiofilm formation - Pseudomonas aeruginosa\n",
      "map02026\tBiofilm formation - Escherichia coli\n",
      "map02030\tBacterial chemotaxis\n",
      "map02040\tFlagellar assembly\n",
      "map04814\tMotor proteins\n",
      "map04820\tCytoskeleton in muscle cells\n",
      "map04810\tRegulation of actin cytoskeleton\n",
      "map04640\tHematopoietic cell lineage\n",
      "map04610\tComplement and coagulation cascades\n",
      "map04611\tPlatelet activation\n",
      "map04613\tNeutrophil extracellular trap formation\n",
      "map04620\tToll-like receptor signaling pathway\n",
      "map04624\tToll and Imd signaling pathway\n",
      "map04621\tNOD-like receptor signaling pathway\n",
      "map04622\tRIG-I-like receptor signaling pathway\n",
      "map04623\tCytosolic DNA-sensing pathway\n",
      "map04625\tC-type lectin receptor signaling pathway\n",
      "map04650\tNatural killer cell mediated cytotoxicity\n",
      "map04612\tAntigen processing and presentation\n",
      "map04660\tT cell receptor signaling pathway\n",
      "map04658\tTh1 and Th2 cell differentiation\n",
      "map04659\tTh17 cell differentiation\n",
      "map04657\tIL-17 signaling pathway\n",
      "map04662\tB cell receptor signaling pathway\n",
      "map04664\tFc epsilon RI signaling pathway\n",
      "map04666\tFc gamma R-mediated phagocytosis\n",
      "map04670\tLeukocyte transendothelial migration\n",
      "map04672\tIntestinal immune network for IgA production\n",
      "map04062\tChemokine signaling pathway\n",
      "map04911\tInsulin secretion\n",
      "map04910\tInsulin signaling pathway\n",
      "map04922\tGlucagon signaling pathway\n",
      "map04923\tRegulation of lipolysis in adipocytes\n",
      "map04920\tAdipocytokine signaling pathway\n",
      "map03320\tPPAR signaling pathway\n",
      "map04929\tGnRH secretion\n",
      "map04912\tGnRH signaling pathway\n",
      "map04913\tOvarian steroidogenesis\n",
      "map04915\tEstrogen signaling pathway\n",
      "map04914\tProgesterone-mediated oocyte maturation\n",
      "map04917\tProlactin signaling pathway\n",
      "map04921\tOxytocin signaling pathway\n",
      "map04926\tRelaxin signaling pathway\n",
      "map04935\tGrowth hormone synthesis, secretion and action\n",
      "map04918\tThyroid hormone synthesis\n",
      "map04919\tThyroid hormone signaling pathway\n",
      "map04928\tParathyroid hormone synthesis, secretion and action\n",
      "map04916\tMelanogenesis\n",
      "map04924\tRenin secretion\n",
      "map04614\tRenin-angiotensin system\n",
      "map04925\tAldosterone synthesis and secretion\n",
      "map04927\tCortisol synthesis and secretion\n",
      "map04260\tCardiac muscle contraction\n",
      "map04261\tAdrenergic signaling in cardiomyocytes\n",
      "map04270\tVascular smooth muscle contraction\n",
      "map04970\tSalivary secretion\n",
      "map04971\tGastric acid secretion\n",
      "map04972\tPancreatic secretion\n",
      "map04976\tBile secretion\n",
      "map04973\tCarbohydrate digestion and absorption\n",
      "map04974\tProtein digestion and absorption\n",
      "map04975\tFat digestion and absorption\n",
      "map04979\tCholesterol metabolism\n",
      "map04977\tVitamin digestion and absorption\n",
      "map04980\tCobalamin transport and metabolism\n",
      "map04978\tMineral absorption\n",
      "map04962\tVasopressin-regulated water reabsorption\n",
      "map04960\tAldosterone-regulated sodium reabsorption\n",
      "map04961\tEndocrine and other factor-regulated calcium reabsorption\n",
      "map04964\tProximal tubule bicarbonate reclamation\n",
      "map04966\tCollecting duct acid secretion\n",
      "map04724\tGlutamatergic synapse\n",
      "map04727\tGABAergic synapse\n",
      "map04725\tCholinergic synapse\n",
      "map04728\tDopaminergic synapse\n",
      "map04726\tSerotonergic synapse\n",
      "map04720\tLong-term potentiation\n",
      "map04730\tLong-term depression\n",
      "map04723\tRetrograde endocannabinoid signaling\n",
      "map04721\tSynaptic vesicle cycle\n",
      "map04722\tNeurotrophin signaling pathway\n",
      "map04744\tPhototransduction\n",
      "map04745\tPhototransduction - fly\n",
      "map04740\tOlfactory transduction\n",
      "map04742\tTaste transduction\n",
      "map04750\tInflammatory mediator regulation of TRP channels\n",
      "map04320\tDorso-ventral axis formation\n",
      "map04360\tAxon guidance\n",
      "map04361\tAxon regeneration\n",
      "map04380\tOsteoclast differentiation\n",
      "map04211\tLongevity regulating pathway\n",
      "map04212\tLongevity regulating pathway - worm\n",
      "map04213\tLongevity regulating pathway - multiple species\n",
      "map04710\tCircadian rhythm\n",
      "map04713\tCircadian entrainment\n",
      "map04711\tCircadian rhythm - fly\n",
      "map04712\tCircadian rhythm - plant\n",
      "map04714\tThermogenesis\n",
      "map04626\tPlant-pathogen interaction\n",
      "map05200\tPathways in cancer\n",
      "map05202\tTranscriptional misregulation in cancer\n",
      "map05206\tMicroRNAs in cancer\n",
      "map05205\tProteoglycans in cancer\n",
      "map05204\tChemical carcinogenesis - DNA adducts\n",
      "map05207\tChemical carcinogenesis - receptor activation\n",
      "map05208\tChemical carcinogenesis - reactive oxygen species\n",
      "map05203\tViral carcinogenesis\n",
      "map05230\tCentral carbon metabolism in cancer\n",
      "map05231\tCholine metabolism in cancer\n",
      "map05235\tPD-L1 expression and PD-1 checkpoint pathway in cancer\n",
      "map05210\tColorectal cancer\n",
      "map05212\tPancreatic cancer\n",
      "map05225\tHepatocellular carcinoma\n",
      "map05226\tGastric cancer\n",
      "map05214\tGlioma\n",
      "map05216\tThyroid cancer\n",
      "map05221\tAcute myeloid leukemia\n",
      "map05220\tChronic myeloid leukemia\n",
      "map05217\tBasal cell carcinoma\n",
      "map05218\tMelanoma\n",
      "map05211\tRenal cell carcinoma\n",
      "map05219\tBladder cancer\n",
      "map05215\tProstate cancer\n",
      "map05213\tEndometrial cancer\n",
      "map05224\tBreast cancer\n",
      "map05222\tSmall cell lung cancer\n",
      "map05223\tNon-small cell lung cancer\n",
      "map05166\tHuman T-cell leukemia virus 1 infection\n",
      "map05170\tHuman immunodeficiency virus 1 infection\n",
      "map05161\tHepatitis B\n",
      "map05160\tHepatitis C\n",
      "map05171\tCoronavirus disease - COVID-19\n",
      "map05164\tInfluenza A\n",
      "map05162\tMeasles\n",
      "map05168\tHerpes simplex virus 1 infection\n",
      "map05163\tHuman cytomegalovirus infection\n",
      "map05167\tKaposi sarcoma-associated herpesvirus infection\n",
      "map05169\tEpstein-Barr virus infection\n",
      "map05165\tHuman papillomavirus infection\n",
      "map05110\tVibrio cholerae infection\n",
      "map05120\tEpithelial cell signaling in Helicobacter pylori infection\n",
      "map05130\tPathogenic Escherichia coli infection\n",
      "map05132\tSalmonella infection\n",
      "map05131\tShigellosis\n",
      "map05135\tYersinia infection\n",
      "map05133\tPertussis\n",
      "map05134\tLegionellosis\n",
      "map05150\tStaphylococcus aureus infection\n",
      "map05152\tTuberculosis\n",
      "map05100\tBacterial invasion of epithelial cells\n",
      "map05146\tAmoebiasis\n",
      "map05144\tMalaria\n",
      "map05145\tToxoplasmosis\n",
      "map05140\tLeishmaniasis\n",
      "map05142\tChagas disease\n",
      "map05143\tAfrican trypanosomiasis\n",
      "map05310\tAsthma\n",
      "map05322\tSystemic lupus erythematosus\n",
      "map05323\tRheumatoid arthritis\n",
      "map05320\tAutoimmune thyroid disease\n",
      "map05321\tInflammatory bowel disease\n",
      "map05330\tAllograft rejection\n",
      "map05332\tGraft-versus-host disease\n",
      "map05340\tPrimary immunodeficiency\n",
      "map05010\tAlzheimer disease\n",
      "map05012\tParkinson disease\n",
      "map05014\tAmyotrophic lateral sclerosis\n",
      "map05016\tHuntington disease\n",
      "map05017\tSpinocerebellar ataxia\n",
      "map05020\tPrion disease\n",
      "map05022\tPathways of neurodegeneration - multiple diseases\n",
      "map05030\tCocaine addiction\n",
      "map05031\tAmphetamine addiction\n",
      "map05032\tMorphine addiction\n",
      "map05033\tNicotine addiction\n",
      "map05034\tAlcoholism\n",
      "map05417\tLipid and atherosclerosis\n",
      "map05418\tFluid shear stress and atherosclerosis\n",
      "map05410\tHypertrophic cardiomyopathy\n",
      "map05412\tArrhythmogenic right ventricular cardiomyopathy\n",
      "map05414\tDilated cardiomyopathy\n",
      "map05415\tDiabetic cardiomyopathy\n",
      "map05416\tViral myocarditis\n",
      "map04930\tType II diabetes mellitus\n",
      "map04940\tType I diabetes mellitus\n",
      "map04950\tMaturity onset diabetes of the young\n",
      "map04936\tAlcoholic liver disease\n",
      "map04932\tNon-alcoholic fatty liver disease\n",
      "map04931\tInsulin resistance\n",
      "map04933\tAGE-RAGE signaling pathway in diabetic complications\n",
      "map04934\tCushing syndrome\n",
      "map01501\tbeta-Lactam resistance\n",
      "map01502\tVancomycin resistance\n",
      "map01503\tCationic antimicrobial peptide (CAMP) resistance\n",
      "map01521\tEGFR tyrosine kinase inhibitor resistance\n",
      "map01524\tPlatinum drug resistance\n",
      "map01523\tAntifolate resistance\n",
      "map01522\tEndocrine resistance\n",
      "map07011\tPenicillins\n",
      "map07012\tCephalosporins - parenteral agents\n",
      "map07013\tCephalosporins - oral agents\n",
      "map07021\tAminoglycosides\n",
      "map07019\tTetracyclines\n",
      "map07020\tMacrolides and ketolides\n",
      "map07014\tQuinolones\n",
      "map07023\tRifamycins\n",
      "map07026\tAntifungal agents\n",
      "map07044\tAntiviral agents\n",
      "map07053\tAnti-HIV agents\n",
      "map07040\tAntineoplastics - alkylating agents\n",
      "map07041\tAntineoplastics - antimetabolic agents\n",
      "map07042\tAntineoplastics - agents from natural products\n",
      "map07043\tAntineoplastics - hormones\n",
      "map07045\tAntineoplastics - protein kinase inhibitors\n",
      "map07032\tHypnotics\n",
      "map07030\tAnxiolytics\n",
      "map07033\tAnticonvulsants\n",
      "map07015\tLocal analgesics\n",
      "map07039\tOpioid analgesics\n",
      "map07028\tAntipsychotics\n",
      "map07029\tAntipsychotics - phenothiazines\n",
      "map07031\tAntipsychotics - butyrophenones\n",
      "map07027\tAntidepressants\n",
      "map07056\tAgents for Alzheimer-type dementia\n",
      "map07057\tAntiparkinsonian agents\n",
      "map07055\tSulfonamide derivatives - overview\n",
      "map07016\tSulfonamide derivatives - sulfa drugs\n",
      "map07017\tSulfonamide derivatives - diuretics\n",
      "map07018\tSulfonamide derivatives - hypoglycemic agents\n",
      "map07037\tAntiarrhythmic drugs\n",
      "map07038\tAntiulcer drugs\n",
      "map07046\tImmunosuppressive agents\n",
      "map07047\tOsteoporosis drugs\n",
      "map07048\tAntimigraines\n",
      "map07049\tAntithrombosis agents\n",
      "map07050\tAntirheumatics - DMARDs and biological agents\n",
      "map07051\tAntidiabetics\n",
      "map07052\tAntidyslipidemic agents\n",
      "map07054\tAntiglaucoma agents\n",
      "map07220\tCholinergic and anticholinergic drugs\n",
      "map07215\talpha-Adrenergic receptor agonists/antagonists\n",
      "map07214\tbeta-Adrenergic receptor agonists/antagonists\n",
      "map07213\tDopamine receptor agonists/antagonists\n",
      "map07212\tHistamine H1 receptor antagonists\n",
      "map07227\tHistamine H2/H3 receptor agonists/antagonists\n",
      "map07211\tSerotonin receptor agonists/antagonists\n",
      "map07228\tEicosanoid receptor agonists/antagonists\n",
      "map07224\tOpioid receptor agonists/antagonists\n",
      "map07229\tAngiotensin receptor and endothelin receptor antagonists\n",
      "map07225\tGlucocorticoid and mineralocorticoid receptor agonists/antagonists\n",
      "map07226\tProgesterone, androgen and estrogen receptor agonists/antagonists\n",
      "map07223\tRetinoic acid receptor (RAR) and retinoid X receptor (RXR) agonists/antagonists\n",
      "map07222\tPeroxisome proliferator-activated receptor (PPAR) agonists\n",
      "map07221\tNicotinic cholinergic receptor antagonists\n",
      "map07230\tGABA-A receptor agonists/antagonists\n",
      "map07036\tCalcium channel blocking drugs\n",
      "map07231\tSodium channel blocking drugs\n",
      "map07232\tPotassium channel blocking and opening drugs\n",
      "map07235\tN-Methyl-D-aspartic acid (NMDA) receptor antagonists\n",
      "map07233\tIon transporter inhibitors\n",
      "map07234\tNeurotransmitter transporter inhibitors\n",
      "map07216\tCatecholamine transferase inhibitors\n",
      "map07219\tCyclooxygenase inhibitors\n",
      "map07024\tHMG-CoA reductase inhibitors\n",
      "map07217\tRenin-angiotensin system inhibitors\n",
      "map07218\tHIV protease inhibitors\n",
      "map07025\tQuinolines\n",
      "map07034\tEicosanoids\n",
      "map07035\tProstaglandins\n",
      "map07110\tBenzoic acid family\n",
      "map07112\t1,2-Diphenyl substitution family\n",
      "map07114\tNaphthalene family\n",
      "map07117\tBenzodiazepine family\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(REST.kegg_list(\"pathway\").read())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We could transform this into a nice table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>map01100</td>\n",
       "      <td>Metabolic pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>map01110</td>\n",
       "      <td>Biosynthesis of secondary metabolites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>map01120</td>\n",
       "      <td>Microbial metabolism in diverse environments</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>map01200</td>\n",
       "      <td>Carbon metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>map01210</td>\n",
       "      <td>2-Oxocarboxylic acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>map07035</td>\n",
       "      <td>Prostaglandins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>map07110</td>\n",
       "      <td>Benzoic acid family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>map07112</td>\n",
       "      <td>1,2-Diphenyl substitution family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>map07114</td>\n",
       "      <td>Naphthalene family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>map07117</td>\n",
       "      <td>Benzodiazepine family</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>571 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Data                                          Name\n",
       "0    map01100                            Metabolic pathways\n",
       "1    map01110         Biosynthesis of secondary metabolites\n",
       "2    map01120  Microbial metabolism in diverse environments\n",
       "3    map01200                             Carbon metabolism\n",
       "4    map01210               2-Oxocarboxylic acid metabolism\n",
       "..        ...                                           ...\n",
       "566  map07035                                Prostaglandins\n",
       "567  map07110                           Benzoic acid family\n",
       "568  map07112              1,2-Diphenyl substitution family\n",
       "569  map07114                            Naphthalene family\n",
       "570  map07117                         Benzodiazepine family\n",
       "\n",
       "[571 rows x 2 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pathways = get_df(REST.kegg_list(\"pathway\"))\n",
    "df_pathways"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And search in it specific pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>map00910</td>\n",
       "      <td>Nitrogen metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Data                 Name\n",
       "32  map00910  Nitrogen metabolism"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pathways[df_pathways[\"Name\"].str.contains(\"Nitrogen\")]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding id within database\n",
    "\n",
    "kegg_find( )\n",
    "\n",
    "With this function you can find entries with matching query keywords or other query data in a specific database.\n",
    "For example, using the same id from above, we could search it in the \"pathway\" database: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "path:map00910\tNitrogen metabolism\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(REST.kegg_find(\"pathway\",\"map00910\").read())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting entry information from id\n",
    "But usually, we want more info about our ids. Then we can use the function kegg_get( )\n",
    "\n",
    "Let's say I want to get more info on the \"Nitrogen metabolism\", we could do it this way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENTRY       map00910                    Pathway\n",
      "NAME        Nitrogen metabolism\n",
      "DESCRIPTION The biological process of the nitrogen cycle is a complex interplay among many microorganisms catalyzing different reactions, where nitrogen is found in various oxidation states ranging from +5 in nitrate to -3 in ammonia. The core nitrogen cycle involves four reduction pathways and two oxidation pathways. Nitrogen fixation [MD:M00175] is the process of reducing atmospheric molecular nitrogen to ammonia, a biologically useful reduced form incorporated into amino acids and other vital compounds. The ability of fixing atmospheric nitrogen by the nitrogenase enzyme complex is present in restricted prokaryotes (diazotrophs). The other reduction pathways are assimilatory nitrate reduction [MD:M00531] and dissimilatory nitrate reduction [MD:M00530] both for conversion to ammonia, and denitrification [MD:M00529]. Denitrification is a respiration in which nitrate or nitrite is reduced as a terminal electron acceptor under low oxygen or anoxic conditions, producing gaseous nitrogen compounds (N2, NO and N2O) to the atmosphere. The two oxidation pathways are nitrification [MD:M00528] and anaerobic ammonium oxidation (anammox) [MD:M00973]. Nitrification is the oxidation of ammonia (NH3) with oxygen into nitrite followed by the oxidation of nitrite into nitrate. The first step is performed by ammonia-oxidizing microorganisms (e.g., Nitrosomonas and Nitrosococcus) and the second step by nitrite-oxidizing microorganisms (e.g., Nitrobacter). Anammox is a biochemical process of oxidizing ammonium (NH4+) into dinitrogen gas (N2) using nitrite as an electron acceptor. It occurs in the anammoxosome, a membrane bound compartment inside the cytoplasm, of anammox bacteria (e.g., Planctomycetes).\n",
      "CLASS       Metabolism; Energy metabolism\n",
      "PATHWAY_MAP map00910  Nitrogen metabolism\n",
      "MODULE      M00175  Nitrogen fixation, nitrogen => ammonia [PATH:map00910]\n",
      "            M00528  Nitrification, ammonia => nitrite [PATH:map00910]\n",
      "            M00529  Denitrification, nitrate => nitrogen [PATH:map00910]\n",
      "            M00530  Dissimilatory nitrate reduction, nitrate => ammonia [PATH:map00910]\n",
      "            M00531  Assimilatory nitrate reduction, nitrate => ammonia [PATH:map00910]\n",
      "            M00804  Complete nitrification, comammox, ammonia => nitrite => nitrate [PATH:map00910]\n",
      "            M00973  Anammox, nitrite + ammonia => nitrogen [PATH:map00910]\n",
      "DBLINKS     GO: 0006807\n",
      "REFERENCE   PMID:14702404\n",
      "  AUTHORS   Scott JD, Ludwig RA\n",
      "  TITLE     Azorhizobium caulinodans electron-transferring flavoprotein N electrochemically couples pyruvate dehydrogenase complex activity to N2 fixation.\n",
      "  JOURNAL   Microbiology 150:117-26 (2004)\n",
      "            DOI:10.1099/mic.0.26603-0\n",
      "REFERENCE   PMID:17408485\n",
      "  AUTHORS   Kneip C, Lockhart P, Voss C, Maier UG\n",
      "  TITLE     Nitrogen fixation in eukaryotes--new models for symbiosis.\n",
      "  JOURNAL   BMC Evol Biol 7:55 (2007)\n",
      "            DOI:10.1186/1471-2148-7-55\n",
      "REFERENCE   PMID:11004450\n",
      "  AUTHORS   Whittaker M, Bergmann D, Arciero D, Hooper AB\n",
      "  TITLE     Electron transfer during the oxidation of ammonia by the chemolithotrophic bacterium Nitrosomonas europaea.\n",
      "  JOURNAL   Biochim Biophys Acta 1459:346-55 (2000)\n",
      "            DOI:10.1016/S0005-2728(00)00171-7\n",
      "REFERENCE   PMID:15528644\n",
      "  AUTHORS   Cabello P, Roldan MD, Moreno-Vivian C\n",
      "  TITLE     Nitrate reduction and the nitrogen cycle in archaea.\n",
      "  JOURNAL   Microbiology 150:3527-46 (2004)\n",
      "            DOI:10.1099/mic.0.27303-0\n",
      "REFERENCE   PMID:11921398\n",
      "  AUTHORS   Stolz JF, Basu P\n",
      "  TITLE     Evolution of nitrate reductase: molecular and structural variations on a common function.\n",
      "  JOURNAL   Chembiochem 3:198-206 (2002)\n",
      "            DOI:10.1002/1439-7633(20020301)3:2/3<198::AID-CBIC198>3.0.CO;2-C\n",
      "REFERENCE   PMID:18021072\n",
      "  AUTHORS   Morozkina EV, Zvyagilskaya RA\n",
      "  TITLE     Nitrate reductases: structure, functions, and effect of stress factors.\n",
      "  JOURNAL   Biochemistry (Mosc) 72:1151-60 (2007)\n",
      "            DOI:10.1134/S0006297907100124\n",
      "REFERENCE   PMID:19247843\n",
      "  AUTHORS   Jetten MS, Niftrik L, Strous M, Kartal B, Keltjens JT, Op den Camp HJ\n",
      "  TITLE     Biochemistry and molecular biology of anammox bacteria.\n",
      "  JOURNAL   Crit Rev Biochem Mol Biol 44:65-84 (2009)\n",
      "            DOI:10.1080/10409230902722783\n",
      "REFERENCE   PMID:22221911\n",
      "  AUTHORS   Luesken FA, Wu ML, Op den Camp HJ, Keltjens JT, Stunnenberg H, Francoijs KJ, Strous M, Jetten MS\n",
      "  TITLE     Effect of oxygen on the anaerobic methanotroph 'Candidatus Methylomirabilis oxyfera': kinetic and transcriptional analysis.\n",
      "  JOURNAL   Environ Microbiol 14:1024-34 (2012)\n",
      "            DOI:10.1111/j.1462-2920.2011.02682.x\n",
      "REFERENCE   PMID:26610024\n",
      "  AUTHORS   Daims H, Lebedeva EV, Pjevac P, Han P, Herbold C, Albertsen M, Jehmlich N, Palatinszky M, Vierheilig J, Bulaev A, Kirkegaard RH, von Bergen M, Rattei T, Bendinger B, Nielsen PH, Wagner M\n",
      "  TITLE     Complete nitrification by Nitrospira bacteria.\n",
      "  JOURNAL   Nature 528:504-9 (2015)\n",
      "            DOI:10.1038/nature16461\n",
      "REFERENCE   PMID:26610025\n",
      "  AUTHORS   van Kessel MA, Speth DR, Albertsen M, Nielsen PH, Op den Camp HJ, Kartal B, Jetten MS, Lucker S\n",
      "  TITLE     Complete nitrification by a single microorganism.\n",
      "  JOURNAL   Nature 528:555-9 (2015)\n",
      "            DOI:10.1038/nature16459\n",
      "REL_PATHWAY map00220  Arginine biosynthesis\n",
      "            map00250  Alanine, aspartate and glutamate metabolism\n",
      "            map00460  Cyanoamino acid metabolism\n",
      "            map00630  Glyoxylate and dicarboxylate metabolism\n",
      "            map00680  Methane metabolism\n",
      "            map00720  Carbon fixation pathways in prokaryotes\n",
      "KO_PATHWAY  ko00910\n",
      "///\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(REST.kegg_get(\"map00910\").read())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see all the modules that are associated with it and recover this information \n",
    "using regex (regular expressions):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ModuleID</th>\n",
       "      <th>Name</th>\n",
       "      <th>mapID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M00175</td>\n",
       "      <td>Nitrogen fixation, nitrogen =&gt; ammonia</td>\n",
       "      <td>map00910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M00528</td>\n",
       "      <td>Nitrification, ammonia =&gt; nitrite</td>\n",
       "      <td>map00910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M00529</td>\n",
       "      <td>Denitrification, nitrate =&gt; nitrogen</td>\n",
       "      <td>map00910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M00530</td>\n",
       "      <td>Dissimilatory nitrate reduction, nitrate =&gt; am...</td>\n",
       "      <td>map00910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M00531</td>\n",
       "      <td>Assimilatory nitrate reduction, nitrate =&gt; amm...</td>\n",
       "      <td>map00910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>M00804</td>\n",
       "      <td>Complete nitrification, comammox, ammonia =&gt; n...</td>\n",
       "      <td>map00910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>M00973</td>\n",
       "      <td>Anammox, nitrite + ammonia =&gt; nitrogen</td>\n",
       "      <td>map00910</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ModuleID                                               Name     mapID\n",
       "0   M00175             Nitrogen fixation, nitrogen => ammonia  map00910\n",
       "1   M00528                  Nitrification, ammonia => nitrite  map00910\n",
       "2   M00529               Denitrification, nitrate => nitrogen  map00910\n",
       "3   M00530  Dissimilatory nitrate reduction, nitrate => am...  map00910\n",
       "4   M00531  Assimilatory nitrate reduction, nitrate => amm...  map00910\n",
       "5   M00804  Complete nitrification, comammox, ammonia => n...  map00910\n",
       "6   M00973             Anammox, nitrite + ammonia => nitrogen  map00910"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pattern = r\"(M0\\d+)\\s+(.*)\\s\\[PATH:(.*)\\]\"\n",
    "\n",
    "modules_info = re.findall(pattern,REST.kegg_get(\"map00910\").read())\n",
    "\n",
    "df_modules = pd.DataFrame(modules_info, columns=[\"ModuleID\", \"Name\", \"mapID\"])\n",
    "df_modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
